### DISCUSSION TITLE
[Feature]: Add native Podman support, maybe let us choose if we want to use Docker or Podman via an env during installation

### DISCUSSION THREAD

**Discussion by peaklabs-dev** [coollabsio/coolify#3137] [Feature Requests]
### Description

Let us choose between docker and podman. Best case would be if we could set a custom environment variable to choose podman or docker during installation.

### Minimal Reproduction (if possible, example repository)

The advantage of Podman would be better security and rootless architecture by default.
-> would help with: https://github.com/coollabsio/coolify/issues/2387

- [ ] Add a Podman environment variable that can be used during installation via the install.sh script and via a docker compose file installation.
- [ ] Add Podman as a drop-in replacement for Docker (or in addition).

### Version

-

**Comment by mpcref** [comment]
In "Podman mode" Coolify could make use of [Quadlet](https://www.redhat.com/sysadmin/quadlet-podman) which comes with Podman to run pods as systemd services.

**Reply by lucasew** [reply]
Also, by extension we could have socket activation on services

Imagine a webapp that is rarely used and it's launched on demand

Svelte apps with node-adapter already support this out of the box

**Comment by harish2704** [comment]
Hello team,
I Just want to inform you that it is possible to run Coolify with Podman. The details are given below.

During the experiment, I found 3 things in Coolify which is not compatible with with a typical Podman setup. They are

1. Coolify has hard dependency on Docker-cli and Docker-compose plugin
    - it is not fully compatible with podman-docker or podman-compose. 
    - So I installed docker-cli from the docker repo without installing Docker service. 
    - This way, docker-cli was communicating to `podman` over `podman.sock`
    - I created a unprivileged user ( say userid=1001 ) and started podman service under that user and created symlink for `docker.sock` 
    - `ln -s /var/run/user/1001/podman/podman.sock /var/run/docker.sock`
2. Coolify is expecting a `Server` with Root access to run containers . 
    - For that, I created a simple systemd based container with `sshd` enabled on same machine ( as unprivileged user )
    - Bind mounted `/var/run/user/1001/podman/podman.sock` ( ie,  /var/run/docker.sock ) to that machine
    - Installed docker-cli on that machine . 
    - This way, docker in that VM is actually communicating to podman running as unprivileged user in the host machine
    - I skipped onboarding after startup and manually configured the `Server` with IP of VM and by setting up ssh key. 
3. While adding a new server, Coolify was checking the version number of docker . This is not compatible with Podman.
    - I think this behavior of Coolify should be fixed and Instead of checking `Server.version` Coolify should check `Server.ApiVersion` for compatibility 
    - I fixed this issue by manually patching ( editing ) the php source code inside the container 
  
  After that, everything was working as expected. I tried deploying couple of applications and it worked. 
  
  In short, No.3 was the  only bug like incompatibility I found during this setup. 
  
Edit: details of  host machine I tried these things

NAME="AlmaLinux"
VERSION="9.3 (Shamrock Pampas Cat)"
ARCH: aarch64

**Comment by harish2704** [comment]
@peaklabs-dev : Can I raise a PR to fix the issue mentioned in No.3 ? 
I mean, i Coolify team agrees that this behavior should be fixed, then I am ready to raise a PR which fixes above mentioned area.

**Comment by harish2704** [comment]
Just sharing one more info
1. Traefik proxy may not work well with Podman due  incompatibility in the way `health_status` event is emitted by Podman. 
2. Podman will emit `health_status` for each health check attempt and each of those attempts will cause Traefik proxy to re-fetch and reload its configuration from docker/podman API.
3. In a server with multiple deployed applications, it is a huge waste of CPU resource . 

I opened a bug report  in podman repo https://github.com/containers/podman/issues/24003

**Comment by peaklabs-dev** [comment]
@harish2704 Thank you for your work. Of course raise a PR and we will check it. We have official plans for podman later down the line once stabel is out but if you are abele to contribute to this now we are always open for contributors.

**Comment by prostori** [comment]
@harish2704 @mpcref @peaklabs-dev @cmer @wutangpaul Are you still running Podman via Coolify? Everything works well? I really want to use Podman instead of a Docker.

**Reply by kpc2** [reply]
Looks like the answer is yes, kind of. He's using [his own patched podman in a non-critical staging environment](https://github.com/containers/podman/pull/24005#issuecomment-2656037785) until his pull request gets merged.

**Comment by AtikoSpeed** [comment]
This.
Could you please add it to Coolify 5 roadmap if it won't be implemented sooner.
@andrasbacsai

**Comment by harish2704** [comment]
Hello All, I have put everything necessary to run Coolify on podman here in this repo 
https://github.com/harish2704/coolify-on-podman 
The script looks like this 
```bash
#!/bin/bash

CURR_USER=$USER;
CURR_UID=$UID;
echo "Need root access for creating /data/coolify directory."
sudo bash -c "mkdir -p /data/coolify; chown $CURR_USER /data/coolify"

echo "Need root access for creating symlink /var/run/docker.sock => /run/user/$CURR_UID/podman/podman.sock"
sudo ln -s /run/user/$CURR_UID/podman/podman.sock /var/run/docker.sock

mkdir -p /data/coolify/{source,ssh,applications,databases,backups,services,proxy,webhooks-during-maintenance}
mkdir -p /data/coolify/ssh/{keys,mux}
mkdir -p /data/coolify/proxy/dynamic

ssh-keygen -f /data/coolify/ssh/keys/id.root@host.docker.internal -t ed25519 -N '' -C root@coolify

cat /data/coolify/ssh/keys/id.root@host.docker.internal.pub >> ./ssh/authorized_keys
chmod 700 ./ssh
chmod 600 ./ssh/authorized_keys

curl -fsSL https://cdn.coollabs.io/coolify/docker-compose.yml -o /data/coolify/source/docker-compose.yml
curl -fsSL https://cdn.coollabs.io/coolify/docker-compose.prod.yml -o /data/coolify/source/docker-compose.prod.yml
curl -fsSL https://cdn.coollabs.io/coolify/.env.production -o /data/coolify/source/.env
curl -fsSL https://cdn.coollabs.io/coolify/upgrade.sh -o /data/coolify/source/upgrade.sh

# The commented code wont work. 9999 is the userid of www-data user inside container.
# in the case of podman, it will be different. Need to do this using podman exec.
# chown -R 9999:root /data/coolify
chmod -R 700 /data/coolify

sed -i "s|APP_ID=.*|APP_ID=$(openssl rand -hex 16)|g" /data/coolify/source/.env
sed -i "s|APP_KEY=.*|APP_KEY=base64:$(openssl rand -base64 32)|g" /data/coolify/source/.env
sed -i "s|DB_PASSWORD=.*|DB_PASSWORD=$(openssl rand -base64 32)|g" /data/coolify/source/.env
sed -i "s|REDIS_PASSWORD=.*|REDIS_PASSWORD=$(openssl rand -base64 32)|g" /data/coolify/source/.env
sed -i "s|PUSHER_APP_ID=.*|PUSHER_APP_ID=$(openssl rand -hex 32)|g" /data/coolify/source/.env
sed -i "s|PUSHER_APP_KEY=.*|PUSHER_APP_KEY=$(openssl rand -hex 32)|g" /data/coolify/source/.env
sed -i "s|PUSHER_APP_SECRET=.*|PUSHER_APP_SECRET=$(openssl rand -hex 32)|g" /data/coolify/source/.env

podman network exists coolify || podman network create coolify


podman compose --env-file /data/coolify/source/.env -f /data/coolify/source/docker-compose.yml -f /data/coolify/source/docker-compose.prod.yml up -d --pull always --remove-orphans --force-recreate


echo "Waiting for 10 seconds to start containers. Some more changes need to be done after that"
sleep 10

echo "Fixing directory ownership"
podman exec -it --user root coolify chown www-data:root -R /var/www/html/storage/app/

echo "Patching coolify source code to bypass podman version check"
podmanVersion=$(podman -v | cut -d ' ' -f3)
podman exec -it  coolify sed -i "s#docker version|head -2|grep -i version#docker version|head -2|grep -i version| sed 's/$podmanVersion/24.0.0/g' #" /var/www/html/app/Livewire/Boarding/Index.php

echo "starting systemd-vm container"
podman compose up -d
```

**Reply by egotei** [reply]
Your code is missing a dash "-" in a critical point cause cannot UP the containers without it:

```bash
# Focus in "--pull always"
podman compose --env-file /data/coolify/source/.env -f /data/coolify/source/docker-compose.yml \ 
-f /data/coolify/source/docker-compose.prod.yml up -d --pull always --remove-orphans --force-recreate

# Should be "--pull-always"
podman compose --env-file /data/coolify/source/.env -f /data/coolify/source/docker-compose.yml \ 
-f /data/coolify/source/docker-compose.prod.yml up -d --pull-always --remove-orphans --force-recreate
```

Also docker-compose.prod.yml has an issue in postgres service on healthcheck:
raise ValueError("'CMD_SHELL' takes a single string after it")
So: 
```bash
      test: [ "CMD-SHELL", "pg_isready -U ${DB_USERNAME}", "-d", "${DB_DATABASE:-coolify}" ]
      
      # Should be
      test: [ "CMD-SHELL", "pg_isready -U ${DB_USERNAME}, -d, ${DB_DATABASE:-coolify}" ]
 ```

**Reply by harish2704** [reply]
Thanks for the update. 
May be that pull option is not standardized across all `compose` plugins. 

I am using following compose plugin 
```
Docker Compose version v2.40.0
```
and in it's help, it is written like this 
```
      --pull string                  Pull image before running ("always"|"missing"|"never") (default "policy")
```
The other issue can also be compose plugin specific issue . I am not fully sure.

**Reply by egotei** [reply]
You are right, I am not expert and I assumed wrongly I use podman-compose so that's my bad.

About the "single string after CMD-SHELL maybe I am wrong as well. I fixed with 2 additional lines after downloading the file but before starting up the containers:
```bash
# I omitted the full path /data/coolify/source thing ...
cat docker-compose.prod.yml | sed 's/", "-d", "/, -d, /' > temp.file 
cp temp.file docker-compose.prod.yml
```

But I wasn't able to make it work in Debian 13 Trixie. Just a white blank website. I guessed that all the 1 click app or service deployment was going to be done with docker so I gave up. I wish it supported native podman ;) 

But thanks for sharing you fix.

**Reply by fercod** [reply]
Can this be used in production?

**Reply by egotei** [reply]
> Can this be used in production?

I don't know what you want to achieve. You could make Coolify run with podman? Sure 

I could not but did not spent much time trying to fix the issue). But will all the services that are installed with just one click in Coolify run with podman? I don't think so. 

So I end up installing the official way as a dev lab just for testing purpose. Let's say I want to try Librechat against OpenWebUI I deploy both of them, test them se which one I like more and destroy both. Then I will go to my production machine, create an Isolated user in the system for that service/app I want to deploy, create a hardened podman-compose.yml (not basic settings) and with the very same script I use to create the user and the environment it creates a systemd script so it starts on every boot. 

100% rootless and more control of what I am doing but more tedious but no user can mess with any other service but it's own. And then I use Pangolin as reverse proxy, manage certs and to connect remote networks. 

If you want to test it is as easy as this:
## Official Quick Installation method  (Recommended)

env ROOT_USERNAME=admin \
ROOT_USER_EMAIL=admin@example.com \
ROOT_USER_PASSWORD=SecurePassword123 \
AUTOUPDATE=false \
DOCKER_ADDRESS_POOL_BASE=172.16.0.0/12 \
REGISTRY_URL=your-registry.com \
bash -c 'curl -fsSL https://cdn.coollabs.io/coolify/install.sh | bash'
